services/QuillLMS/app/controllers/concerns/results_summary.rb

```diff
@@ -0,0 +1,117 @@
+module ResultsSummary
+  extend ActiveSupport::Concern
+
+  NOT_PRESENT = 'Not present'
+  NOT_CORRECT = 'Not correct'
+  FULLY_CORRECT = 'Fully correct'
+  PARTIALLY_CORRECT = 'Partially correct'
+
+  NO_PROFICIENCY = 'No proficiency'
+  PARTIAL_PROFICIENCY = 'Partial proficiency'
+  PROFICIENCY = 'Proficiency'
+
+  extend self
+
+  def results_summary(current_user, activity_id, classroom_id, unit_id)
+    @current_user = current_user
+    activity = Activity.find(activity_id)
+    @skill_groups = activity.skill_groups
+    set_activity_sessions_and_assigned_students(activity_id, classroom_id, unit_id)
+    @skill_group_summaries = @skill_groups.map do |skill_group|
+      {
+        name: skill_group.name,
+        description: skill_group.description,
+        not_yet_proficient_student_names: []
+      }
+    end
+
+    {
+      student_results: student_results,
+      skill_group_summaries: @skill_group_summaries
+    }
+  end
+
+  private def set_activity_sessions_and_assigned_students(activity_id, classroom_id, unit_id)
+    if unit_id
+      classroom_unit = ClassroomUnit.find_by(unit_id: unit_id, classroom_id: classroom_id)
+      @assigned_students = User.where(id: classroom_unit.assigned_student_ids).sort_by { |u| u.last_name }
+      @activity_sessions = ActivitySession.where(classroom_unit: classroom_unit, state: 'finished')
+    else
+      unit_ids = @current_user.units.joins("JOIN unit_activities ON unit_activities.activity_id = #{activity_id}")
+      classroom_units = ClassroomUnit.where(unit_id: unit_ids, classroom_id: classroom_id)
+      assigned_student_ids = classroom_units.map { |cu| cu.assigned_student_ids }.flatten.uniq
+      @assigned_students = User.where(id: assigned_student_ids).sort_by { |u| u.last_name }
+      @activity_sessions = ActivitySession.where(activity_id: activity_id, classroom_unit_id: classroom_units.ids, state: 'finished').order(completed_at: :desc).uniq { |activity_session| activity_session.user_id }
+    end
+  end
+
+  private def student_results
+    @assigned_students.map do |assigned_student|
+      activity_session = @activity_sessions.find { |as| as.user_id == assigned_student.id }
```

Comment:
Same note as above about this being a faster look-up if `@activity_sessions` is a hash instead of an array.